app/controllers/api/mixins/genealogy.rb

```diff
@@ -0,0 +1,51 @@
+module Api
+  module Mixins
+    module Genealogy
+      RELATIONSHIP_COLLECTIONS = %w[vms templates].freeze
+      VALID_EDIT_ATTRS = %w[description name child_resources parent_resource].freeze
+
+      def edit_resource_with_genealogy(type, id, data)
+        attrs = validate_edit_data(data)
+        parent, children = build_parent_children(data)
+        resource_search(id, type).tap do |resource|
+          resource.replace_children(children)
+          resource.set_parent(parent)
+          resource.update!(attrs)
+        end
```

Comment:
in the followup, think this is the part that is genealogy. adding the children.
having the concept of valid params and the like seems part of the controller
having the concepts of adding children seems part of the mixin